06_gpu_and_ml/text-to-video/mochi.py

```diff
@@ -28,338 +27,147 @@
+# ## Setting up our Mochi class
+# We'll use the `modal.cls` decorator to define a [Modal Class](https://modal.com/docs/guide/lifecycle-functions)
+# which we use to control the lifecycle of our cloud container.
+# 
+# We configure it to use our image, the distributed volume, and a single H100 GPU.
 
-    snapshot_download(
-        repo_id="genmo/mochi-1-preview",
-        local_dir=MODEL_CACHE / "mochi-1-preview",
-        revision=model_revision,
+with image.imports():
+    import torch
+    from diffusers import MochiPipeline
+    from diffusers.utils import export_to_video
+    
+@app.cls(
+    image=image,
+    volumes={
+        OUTPUTS_PATH: outputs,  # videos will be saved to a distributed volume
+    },
+    gpu=modal.gpu.H100(count=1),
+    timeout= 1 * HOURS,
     )
-    print("🍡 model downloaded")
-
-    print("🍡 downloading text encoder")
-
-    T5Tokenizer.from_pretrained("google/t5-v1_1-xxl", legacy=False)
-    T5EncoderModel.from_pretrained("google/t5-v1_1-xxl")
-
-    model.commit()
-    print("🍡 text encoder downloaded")
+class Mochi:
+    # We can stack the `modal.build` and `modal.enter` decorators
+    # because we use the diffusers library, which caches the model to disk on its first run.
+    # This builds our model weights into the container image, ready for future use.
+    @modal.build()
+    @modal.enter()
+    def load_model(self):
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        self.pipe = MochiPipeline.from_pretrained(
+            "genmo/mochi-1-preview", 
+            torch_dtype=torch.bfloat16, 
+        )
+        self.pipe.enable_model_cpu_offload()
+        self.pipe.enable_vae_tiling()
 
+    @modal.method()
+    def generate(self, 
+        prompt, 
+        negative_prompt = "",
+        num_inference_steps = 200,
+        guidance_scale = 4.5,
+        num_frames = 19,
+    ):
+        frames = self.pipe(
+            prompt=prompt, 
+            negative_prompt=negative_prompt,
+            num_inference_steps=num_inference_steps, 
+            guidance_scale=guidance_scale,
+            num_frames=num_frames,
+        ).frames[0]
+        
+        # save to disk using prompt as filename
+        for char in string.punctuation:
+            prompt = prompt.replace(char, "")
+        prompt = prompt.replace(" ", "_")
+        prompt = prompt[:230] # since filenames can't be longer than 255 characters
+        mp4_name = str(int(time.time())) + "_" + prompt + ".mp4"
+        
+        export_to_video(frames, Path(OUTPUTS_PATH) /  mp4_name)
+        outputs.commit()
+        return mp4_name
 
 # ## Running Mochi inference
 
 # We can trigger Mochi inference from our local machine by running the code in
 # the local entrypoint below.
 
-# It ensures the model is downloaded to a remote volume,
-# spins up a new replica to generate a video, also saved remotely,
-# and then downloads the video to the local machine.
+# It ensures the model is downloaded into the image,
```

Comment:
Nit: I wouldn't include this language if the model is in the image. It's not the `local_entrypoint`'s responsibility to build an image. This language was present because of the call to `download_model.remote`.

Assuming we add back in the model download call, we'd revert to the original text.